src/backend/utils/adt/network.c

```diff
@@ -389,6 +412,413 @@ network_cmp(PG_FUNCTION_ARGS)
+ * | 1 bit IP |         63 bits netmask         |    (up to 65 bits
+ * |  family  |           (truncated)           |   netmask omitted)
+ * +----------+---------------------------------+
+ *
+ */
+static Datum
+network_abbrev_convert(Datum original, SortSupport ssup)
+{
+	network_sortsupport_state *uss = ssup->ssup_extra;
+	inet	   *authoritative = DatumGetInetPP(original);
+	Datum		res;
+	Datum		ipaddr_datum,
+				netmask,
+				subnet,
+				subnet_bitmask;
+	int			datum_subnet_size;
+
+	/*
+	 * If another IP family is ever added, we'll need to redesign the key
+	 * abbreviation strategy.
+	 */
+	Assert(ip_family(authoritative) == PGSQL_AF_INET ||
+		   ip_family(authoritative) == PGSQL_AF_INET6);
+
+	res = (Datum) 0;
+	if (ip_family(authoritative) == PGSQL_AF_INET6)
+	{
+		/* Shift a 1 over to the datum's most significant bit. */
+		res = ((Datum) 1) << (SIZEOF_DATUM * BITS_PER_BYTE - 1);
+	}
+
+	/*
+	 * Create an integer representation of the IP address by taking its first
+	 * 4 or 8 bytes. We take 8 bytes of an IPv6 address on a 64-bit machine
+	 * and 4 bytes on a 32-bit. Always take all 4 bytes of an IPv4 address.
+	 *
+	 * We're consuming an array of char, so make sure to byteswap on little
+	 * endian systems (an inet's IP array emulates big endian in that the
+	 * first byte is always the most significant).
+	 */
+	if (ip_family(authoritative) == PGSQL_AF_INET6)
+	{
+		ipaddr_datum = *((Datum *) ip_addr(authoritative));
+		ipaddr_datum = DatumBigEndianToNative(ipaddr_datum);
+	}
+	else
+	{
+		uint32		ipaddr_datum32 = *((uint32 *) ip_addr(authoritative));
+#ifndef WORDS_BIGENDIAN
+		ipaddr_datum = pg_bswap32(ipaddr_datum32);
+#endif
+	}
+
+	/*
+	 * Number of bits in subnet. e.g. An IPv4 that's /24 is 32 - 24 = 8.
+	 *
+	 * However, only some of the bits may have made it into the fixed sized
+	 * datum, so take the smallest number between bits in the subnet and bits
+	 * in the datum which are not part of the netmask.
+	 */
+	datum_subnet_size = Min(ip_maxbits(authoritative) - ip_bits(authoritative),
+							SIZEOF_DATUM * BITS_PER_BYTE - ip_bits(authoritative));
+
+	/* we may have ended up with < 0 for a large netmask */
+	if (datum_subnet_size <= 0)
+	{
+		/* the netmask occupies the entirety of datum */
+		subnet = (Datum) 0;
+		netmask = ipaddr_datum;
+	}
+	else
+	{
+		Assert(datum_subnet_size <= SIZEOF_DATUM * BITS_PER_BYTE);
+
+		/*
+		 * This shift creates a power of two like `0010 0000`, and subtracts
+		 * one to create a bitmask for an IP's subnet bits like `0001 1111`.
+		 *
+		 * Note that `datum_subnet_mask` may be == 0, in which case we'll
+		 * generate a 0 bitmask and `subnet` will also come out as 0.
+		 */
+		subnet_bitmask = (((Datum) 1) << datum_subnet_size) - 1;
+
+		/* bitwise AND the IP and bitmask to extract just the subnet bits */
+		subnet = ipaddr_datum & subnet_bitmask;
+
+		/* and likewise, use the mask's complement to get the netmask bits */
+		netmask = ipaddr_datum & ~subnet_bitmask;
```

Comment:
Why not do this before subnet is initialized, in the order that they'll be used in the final abb key?

Reply:
By this do you just mean flip the two lines around? So like this:

``` c
		/* and likewise, use the mask's complement to get the netmask bits */
		netmask = ipaddr_datum & ~subnet_bitmask;

		/* bitwise AND the IP and bitmask to extract just the subnet bits */
		subnet = ipaddr_datum & subnet_bitmask;
```

Reply:
yes

Reply:
👍 Done!